### PR TITLE
Add code for Resync of list of devices with force flag and fix PR com…

### PR DESCRIPTION
**Pull request by Abhishek-121**
…ments of adding different device types

### CODE REVIEW

File: plugins/modules/inventory_intent.py [hunk 1]
```diff
@@ -594,11 +579,46 @@ def get_device_params(self, params):
...
+
+                if response:
+                    log(str(response))
+                    response = response.get("response")[0]
+                    device_id = response["id"]
+                    device_ids.append(device_id)
+
+            except Exception as e:
+                log("An error occurred while fetching the device from Cisco DNA Center")
+                raise Exception("Error while fetching device from Cisco DNA Center")
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:618]: can we print e as well? something like:
raise Exception("Error while fetching device from Cisco DNA Center") from e
(or)
raise Exception(f"'Error while fetching device from Cisco DNA Center' - {str(e)}")

File: plugins/modules/inventory_intent.py [hunk 2]
```diff
@@ -639,12 +659,55 @@ def get_diff_merged(self, config):
...
+                    log("Device Resynced Successfully")
+                    log("Resynced devices are :" + str(device_ips))
+                    msg = "Device " + str(device_ips) + " Resynced Successfully !!"
+                    self.result['log'].append(msg)
+
+                    return self
+
+            except Exception as e:
+                log("An error occurred while Resyncing device in Cisco DNA Center")
+                raise Exception("Error while Resyncing device in Cisco DNA Center")
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:709]: print e

File: plugins/modules/inventory_intent.py [hunk 3]
```diff
@@ -639,12 +659,55 @@ def get_diff_merged(self, config):
...
+                    'payload': device_ids,
+                    'force_sync': force_sync
+                }
+                response = self.dnac._exec(
+                    family="devices",
+                    function='sync_devices_using_forcesync',
+                    op_modifies=True,
+                    params=resync_param_dict,
+                )
+                log(str(response))
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:684]: self.log

File: plugins/modules/inventory_intent.py [hunk 4]
```diff
@@ -639,12 +659,55 @@ def get_diff_merged(self, config):
 
         device_added = False
         device_updated = False
-        # device_resynced = False
 
         devices_to_add = self.have["device_not_in_dnac"]
-        device_type = self.config[0].get('type')
+        device_type = self.config[0].get("type", "NETWORK_DEVICE")
+        device_resynced = self.config[0].get("device_resync", "False")
         self.result['log'] = []
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:667]: Can we add notes above resync in the API docstring description or comments here?

triggers the resync operation using the retrieved device IDs and force sync parameter.

File: plugins/modules/inventory_intent.py [hunk 5]
```diff
@@ -639,12 +659,55 @@ def get_diff_merged(self, config):
...
+                        if 'Synced' in execution_details.get("progress"):
+                            self.status = "success"
+                            self.result['changed'] = True
+                            self.result['response'] = execution_details
+                            break
+                        elif execution_details.get("isError") and execution_details.get("failureReason"):
+                            self.msg = "Device Resynced get failed because of {0}".format(execution_details.get("failureReason"))
+                            self.status = "failed"
+                            break
+                    log("Device Resynced Successfully")
```
Comment (madhansansel) [inline plugins/modules/inventory_intent.py:700]: self.log next line as well

File: plugins/modules/inventory_intent.py [hunk 6]
```diff
@@ -297,6 +297,22 @@
...
+    dnac_username: "{{dnac_username}}"
+    dnac_password: "{{dnac_password}}"
+    dnac_verify: "{{dnac_verify}}"
+    dnac_port: "{{dnac_port}}"
+    dnac_version: "{{dnac_version}}"
+    dnac_debug: "{{dnac_debug}}"
+    dnac_log: False
+    state: merged
+    config:
+      - ipAddress: string
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:312]: I followed camelCase approach(ipAddress/forceSynce) as it is followed in requested parameter in DNAC as well.

File: plugins/modules/inventory_intent.py [hunk 7]
```diff
@@ -594,11 +579,46 @@ def get_device_params(self, params):
...
+        for device_ip in device_ips:
+            try:
+                response = self.dnac._exec(
+                    family="devices",
+                    function='get_device_list',
+                    params={"managementIpAddress": device_ip}
+                )
+
+                if response:
+                    log(str(response))
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:611]: changed.

File: plugins/modules/inventory_intent.py [hunk 8]
```diff
@@ -594,11 +579,46 @@ def get_device_params(self, params):
...
         return device_param
 
+    def get_device_ids(self, device_ips):
+        """
+        Get the list of unique device ID's for list of specified management IP addresses of devices in Cisco DNA Center.
+        Args:
+            self (object): An instance of a class used for interacting with Cisco DNA Center.
+            device_ips (list): The management IP addresses of devices for which you want to retrieve the device ID's.
+        Returns:
+            list: The list of unique device ID's for the specified devices.
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:595]: addressed.

File: plugins/modules/inventory_intent.py [hunk 9]
```diff
@@ -594,11 +579,46 @@ def get_device_params(self, params):
...
-            "update_management_ip_list": params.get("updateMgmtIPaddressList")
+            "update_management_ip_list": params.get("updateMgmtIPaddressList"),
+            "force_sync": params.get("force_sync")
         }
 
         return device_param
 
+    def get_device_ids(self, device_ips):
+        """
+        Get the list of unique device ID's for list of specified management IP addresses of devices in Cisco DNA Center.
```
[the remark this answers is not in the document]
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:590]: addressed

File: plugins/modules/inventory_intent.py (hunk 4 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py:667]: added in the comments.

File: plugins/modules/inventory_intent.py (hunk 5 above)
Comment (Abhishek-121) [reply plugins/modules/inventory_intent.py]: addressed.

**Review by madhansansel — APPROVED** [review]

**Pull request merged**